### PR TITLE
Double range of Timestamp

### PR DESCRIPTION
**Pull request by tgulacsi**

### CODE REVIEW

File: ulid.go [hunk 1]
```diff
@@ -283,10 +283,10 @@ func Now() uint64 { return Timestamp(time.Now().UTC()) }
 // Timestamp converts a time.Time to Unix milliseconds.
 //
 // BUG(tsenart): Because of the way time.Time is internally
-// represented, times from the year 2262 on have undefined
+// represented, times from the year 4524 on have undefined
```
Comment (tsenart) [inline ulid.go]: Are you sure about this? Could we perhaps write a test for such?

**Comment by tgulacsi** [comment]
The test shows that MaxTime can be represented, too.

**Comment by tsenart** [comment]
Nice! Does this test fail for the previous implementation?

**Comment by tgulacsi** [comment]
Yes:
$ git checkout 70540e95fd5e9f833503e01c96e6ec8368d41122 ulid.go
:gthomas@waterhouse: ~/src/github.com/oklog/ulid
$ go test -run=Timestamp
--- FAIL: TestTimestamp (0.00s)
        ulid_test.go:284: MaxTime=281474976710655 = 10889-08-02 06:31:50.655 +0100 CET
        ulid_test.go:286: ts=4773815605011
        ulid_test.go:288: Too big difference in MaxTime and Timestamp(Unix(MaxTime)): mt=281474976710655 = 10889-08-02 06:31:50.655 +0100 CET != 4773815605011

File: ulid.go [hunk 2]
```diff
@@ -276,17 +276,19 @@ var maxTime = ULID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}.Time()
...
 //   Timestamp(time.Now().UTC())
 func Now() uint64 { return Timestamp(time.Now().UTC()) }
 
 // Timestamp converts a time.Time to Unix milliseconds.
 //
-// BUG(tsenart): Because of the way time.Time is internally
-// represented, times from the year 2262 on have undefined
-// results.
+// Because of the way ULID stores time, times from the year
+// 10889 has undefined results.
```
Comment (tsenart) [inline ulid.go]: `produces` has a better ring to it I think.

File: ulid.go [hunk 3]
```diff
@@ -276,17 +276,19 @@ var maxTime = ULID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}.Time()
...
 func Timestamp(t time.Time) uint64 {
-	return uint64(t.Truncate(time.Millisecond).UnixNano() / int64(time.Millisecond))
+	// t.Unix() returns seconds since epoch
+	return uint64(t.Unix())*1000 +
+		// t.NanoSecond() returns the nanoseconds in the last second.
+		uint64(t.Nanosecond()/int(time.Millisecond))
```
Comment (tsenart) [inline ulid.go]: These comments don't add too much value I think. I'd rather jump to documentation when needed.

File: ulid_test.go [hunk 4]
```diff
@@ -278,6 +278,15 @@ func TestTimestamp(t *testing.T) {
 	if got, want := ulid.Timestamp(tm), uint64(1000); got != want {
 		t.Errorf("for %v, got %v, want %v", tm, got, want)
 	}
+
+	mt := ulid.MaxTime()
+	dt := time.Unix(int64(mt/1000), int64((mt%1000)*1000000)).Truncate(time.Millisecond)
+	t.Logf("MaxTime=%d = %v", mt, dt)
+	ts := ulid.Timestamp(dt)
+	t.Logf("ts=%d", ts)
+	if ts-mt > 1000 || mt-ts > 1000 {
```
Comment (tsenart) [inline ulid_test.go]: Why a test of distance? We should be testing equality as far as I can see.

Comment (tsenart) [inline ulid_test.go]: With that, I'd remove the above logging calls and add the following, keeping it consistent with the rest of the tests.

```go
if got, want := ts, mt; got != want {
    t.Errorf("got timestamp %d, want %d", got, want)
}
```

Comment (tgulacsi) [inline ulid_test.go]: Done, thanks.

File: ulid.go (hunk 2 above)
Comment (tgulacsi) [inline ulid.go]: Done.

**Pull request merged**